Add mass SPICE replacement

martinpitt opened the pull request:
When opening the SPICE dialog from the overview page and there are other
SPICE VMs, offer them for conversion as well, with a menu. Reconfigure
the dialog title and verbiage accordingly.

To make some room in the dialog, move the details of what happens into a popover.

----

[pixel diff review](https://github.com/cockpit-project/pixel-test-reference/compare/e47ae65d70b617aa594aed3964dcd948da668dc4..04230218867ee5dbd413ed03d8474617ec332f70)

This implements @garrett's [remaining design for this feature](https://github.com/cockpit-project/cockpit-machines/pull/1387#issuecomment-1895353645).

## Machines: Mass SPICE conversion

The "Replace SPICE devices" action got introduced in the [previous version](https://cockpit-project.org/blog/cockpit-310.html) to convert SPICE-using VMs to run on a host without SPICE capability. When opened from the overview, that dialog now offers you to convert multiple or all VMs in a single step.

![Screen Shot 2024-02-08 at 16 36 00](https://github.com/cockpit-project/cockpit-machines/assets/200109/ebbed4df-3ffc-469e-bf26-670a0308dec9)

cockpituous commented on src/components/vm/vmReplaceSpiceDialog.jsx:
```diff
@@ -32,15 +35,24 @@ import { domainReplaceSpice } from '../../libvirtApi/domain.js';
+export const ReplaceSpiceDialog = ({ vm, vms }) => {
+    const spiceVMs = vms?.filter(vm => vm.inactiveXML?.hasSpice);
+    // sort spicVMs by name, put the current VM first
+    spiceVMs?.sort((a, b) => (a.name === vm.name) ? -1 : (b.name === vm.name) ? 1 : a.name.localeCompare(b.name));
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240201-163147-ce7caf1d-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240203-041307-a2627c26-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240203-084715-a2627c26-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240205-054255-d45075bd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240205-065324-d45075bd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/vm/vmReplaceSpiceDialog.jsx:
```diff
@@ -58,9 +70,44 @@
+        const onSelect = (_ev, item) => setSelected(
+            selected.includes(item)
+                ? selected.filter(id => id !== item)
+                : [...selected, item]
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240201-163147-ce7caf1d-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240203-041307-a2627c26-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240203-084715-a2627c26-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240205-054255-d45075bd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/vm/vmReplaceSpiceDialog.jsx:
```diff
@@ -58,9 +70,44 @@ export const ReplaceSpiceDialog = ({ vmName, vmId, connectionName, vmRunning })
+        const onSelect = (_ev, item) => setSelected(
+            selected.includes(item)
+                ? selected.filter(id => id !== item)
+                : [...selected, item]
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240205-065324-d45075bd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

martinpitt commented:
@garrett Could you please do a first review of the screenshot only? Once you are happy, I'll add a pixel test here, but given how fast they get out of date, I'd rather do that as the last step.

garrett reviewed (changes requested):
@martinpitt: Apologies. I had the review ongoing for several hours today. I needed to respond with a mockup, and kept getting sidetracked and was in a few meetings since too.

Here's how it should probably be:

![image](https://github.com/cockpit-project/cockpit-machines/assets/10246/e976c5be-12ec-4619-900d-ffa9838103bb)

1. Move the details into a popover.
2. Move the alert into helper text, as it's proactive message instead of a reactive one.

I also had a mockup with the select all as a footer, but it looks a bit odd without a separator. But the menu has a separator with footers, so I think that might work?

![image](https://github.com/cockpit-project/cockpit-machines/assets/10246/8acc76ae-c21a-4157-8dd6-28d13b7da0b4)

And do we want "Deselect others" or "Deselect other VMs" instead of "Select none"?

![image](https://github.com/cockpit-project/cockpit-machines/assets/10246/ae4ccd64-8266-41d5-a687-5b043365150a)


(On the topic of "Deselect" vs "Unselect": https://english.stackexchange.com/questions/18465/unselect-or-deselect)

martinpitt commented:
@garrett : I moved the details to the popover, did the "Deselect others" behaviour/string/layout change, and changed the verbiage:

![image](https://github.com/cockpit-project/cockpit-machines/assets/200109/e8400663-b8e9-49f2-a410-7f36e8677996)

It's now very close, except for turning the alert into a helper text. This is a bit fiddly and needs to happen across all dialogs, so if it's ok with you I'd rather do this in a separate PR?

garrett commented:
The fix for the blurred line:

```scss
.pf-v5-c-menu.pf-m-scrollable {
    --pf-v5-c-menu__footer--BoxShadow: 0 -1px var(--pf-v5-global--BorderColor--100);
}
```

This changes the blurry shadow (which looks like a bug) for a sharp 1px border.

(We might even want to do this in PF overrides, if we do something similar as to this.)

garrett commented on src/components/vm/vmReplaceSpiceDialog.jsx:
```diff
@@ -74,16 +123,21 @@ export const ReplaceSpiceDialog = ({ vmName, vmId, connectionName, vmRunning })
+                        : _("Reconfigure the virtual machine for a host which does not support SPICE, for host upgrades or live migration:")
+                }</Text>
                 <TextList>
                     <TextListItem>{_("Convert SPICE graphics console to VNC.")}</TextListItem>
                     <TextListItem>{_("Convert QXL video card to VGA.")}</TextListItem>
                     <TextListItem>{_("Remove SPICE audio and host devices.")}</TextListItem>
```
Nitpick: For lists with fragment phrases, you don't use periods, unless there are two separate thoughts, then you'd use some kind of punctuation to separate them and apply it to all (to stay consistent).

For this list, we shouldn't use periods for any of them.

```suggestion
                    <TextListItem>{_("Convert SPICE graphics console to VNC")}</TextListItem>
                    <TextListItem>{_("Convert QXL video card to VGA")}</TextListItem>
                    <TextListItem>{_("Remove SPICE audio and host devices")}</TextListItem>
```

PatternFly doesn't touch on this, except for headings and titles @ https://www.patternfly.org/ux-writing/punctuation — but it's a general thing when writing, even outside of computers. https://proofed.com/writing-tips/should-bullet-point-lists-end-with-a-period/

It's technically not wrong, as-is, but it's still best to leave out periods for UIs, unless needed.

martinpitt replied:
Done, thanks!

garrett reviewed (dismissed):
Looks great! Thanks.

One very minor thing: The list has periods, and it shouldn't unless needed. (I'm glad it's at least self-consistent!)

martinpitt commented:
Popover:

![image](https://github.com/cockpit-project/cockpit-machines/assets/200109/8a181803-905d-4951-92bf-22a793a6d09b)

I'll update the screenshot in the description for the "closed popover" state.

cockpituous commented on src/components/vm/vmReplaceSpiceDialog.jsx:
```diff
@@ -30,17 +35,28 @@ import { fmt_to_fragments } from 'utils.jsx';
+export const ReplaceSpiceDialog = ({ vm, vms }) => {
+    const spiceVMs = vms?.filter(vm => vm.inactiveXML?.hasSpice);
+    // sort spicVMs by name, put the current VM first
+    spiceVMs?.sort((a, b) => (a.name === vm.name) ? -1 : (b.name === vm.name) ? 1 : a.name.localeCompare(b.name));
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240208-145413-12ec3af2-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/vm/vmReplaceSpiceDialog.jsx:
```diff
@@ -58,9 +74,47 @@
+        const onSelect = (_ev, item) => setSelected(
+            selected.includes(item)
+                ? selected.filter(id => id !== item)
+                : [...selected, item]
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240208-145413-12ec3af2-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

martinpitt commented:
Uh, [this is relevant](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240208-145413-12ec3af2-fedora-39-devel/log.html#85) -- PF doesn't like trying to put a TextList into a Popover?

garrett reviewed (dismissed):
Thanks!

cockpituous commented on src/components/vm/vmReplaceSpiceDialog.jsx:
```diff
@@ -30,17 +36,28 @@ import { fmt_to_fragments } from 'utils.jsx';
+export const ReplaceSpiceDialog = ({ vm, vms }) => {
+    const spiceVMs = vms?.filter(vm => vm.inactiveXML?.hasSpice);
+    // sort spicVMs by name, put the current VM first
+    spiceVMs?.sort((a, b) => (a.name === vm.name) ? -1 : (b.name === vm.name) ? 1 : a.name.localeCompare(b.name));
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240208-153345-081e188a-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240208-171835-081e188a-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/vm/vmReplaceSpiceDialog.jsx:
```diff
@@ -58,9 +75,45 @@ export const ReplaceSpiceDialog = ({ vmName, vmId, connectionName, vmRunning })
+        const onSelect = (_ev, item) => setSelected(
+            selected.includes(item)
+                ? selected.filter(id => id !== item)
+                : [...selected, item]
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240208-153345-081e188a-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240208-171835-081e188a-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

mvollmer commented on src/components/vm/vmReplaceSpiceDialog.jsx:
```diff
@@ -30,17 +36,28 @@ import { fmt_to_fragments } from 'utils.jsx';
     const Dialogs = useDialogs();
     const [error, dialogErrorSet] = useState(null);
     const [inProgress, setInProgress] = useState(false);
+    const [selected, setSelected] = useState(isMultiple ? [spiceVMs[0].id] : []);
```
It's quite confusing how the "primary" VM (passed in `vm`) is handled re the various lists of vms. The primary is part of `spiceVms`, and might be part of `selected` (but only if `isMultiple` is true), so it might also be part of `selectedVMs` (in the multi case), and wouldn't it then be converted twice?

martinpitt replied:
Right, the dialog needs to handle both cases (single and multi-VM) separately. Thanks for spotting!

mvollmer commented on src/components/vm/vmReplaceSpiceDialog.jsx:
```diff
@@ -30,17 +36,28 @@ import { fmt_to_fragments } from 'utils.jsx';
 
+        const selectedVMs = spiceVMs?.filter(vm => selected.includes(vm.id)) || [];
+
         return domainReplaceSpice({ connectionName: vm.connectionName, id: vm.id })
```
So the primary VM always gets replaced, eventhough there is a checkbox for it in the multi case that the user might uncheck?

martinpitt replied:
Indeed, thanks for spotting! I created an integration test for covering this case. The dialog is rather complicated due to the possibility of unchecking the VM which you triggered the action on :cry:

mvollmer commented on src/components/vm/vmReplaceSpiceDialog.jsx:
```diff
@@ -30,17 +36,28 @@ import { fmt_to_fragments } from 'utils.jsx';
+        const selectedVMs = spiceVMs?.filter(vm => selected.includes(vm.id)) || [];
+
         return domainReplaceSpice({ connectionName: vm.connectionName, id: vm.id })
+                .then(() => Promise.allSettled(selectedVMs.map(vm => domainReplaceSpice({ connectionName: vm.connectionName, id: vm.id }))))
```
This converts all VMs in parallel, which is fine for Cockpit and pipelining over slow connections, but might be stressing poor libvirt. Would it make sense to be conservative and convert them one-by-one? Or do we know that libvirt is fine with this?

martinpitt replied:
The conversion is not that expensive, but let's check. My VM can host about 60 running micro-VMs until it runs out of memory:
```
for i in `seq 60`; do virt-install --memory 50 --pxe --virt-type qemu --os-variant alpinelinux3.8 --disk none --wait 0 --name test$i; done
```
This takes about 5 seconds and succeeds. I'll try again with serial conversion, if that doesn't take an undue amount of time, we can play it safe and do it serially.

mvollmer replied:
> This takes about 5 seconds and succeeds.

The conversion takes 5 seconds, right? (Not the creation of the VMs.)

martinpitt replied:
Right, I meant the "Replace" button. Creating the VMs takes much longer.

mvollmer reviewed (commented):
The code was a bit hard to follow, and I think the primary VM will be replaced twice (without any ill effect, probably). Please check that and fix if you think it is a good idea.

cockpituous commented on src/components/vm/vmReplaceSpiceDialog.jsx:
```diff
@@ -30,42 +36,95 @@ import { fmt_to_fragments } from 'utils.jsx';
+        const onSelect = (_ev, item) => setSelected(
+            selected.includes(item)
+                ? selected.filter(id => id !== item)
+                : [...selected, item]
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240209-102132-d1b11d6f-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

martinpitt commented on src/components/vm/vmReplaceSpiceDialog.jsx:
```diff
@@ -30,42 +36,95 @@ import { fmt_to_fragments } from 'utils.jsx';
+export const ReplaceSpiceDialog = ({ vm, vms }) => {
+    const spiceVMs = vms?.filter(vm => vm.inactiveXML?.hasSpice);
+    // sort spiceVMs by name, put the current VM first
+    spiceVMs?.sort((a, b) => (a.name === vm.name) ? -1 : (b.name === vm.name) ? 1 : a.name.localeCompare(b.name));
```
They are guaranteed *not* to be unique across different connections; you can easily have a "test1" on system and session. Indeed this could be `a === vm` and `b === vm`, I'll test if that works.

martinpitt commented on src/components/vm/vmReplaceSpiceDialog.jsx:
```diff
@@ -30,42 +36,95 @@ import { fmt_to_fragments } from 'utils.jsx';
     const Dialogs = useDialogs();
     const [error, dialogErrorSet] = useState(null);
     const [inProgress, setInProgress] = useState(false);
+    const [selected, setSelected] = useState(isMultiple ? [spiceVMs[0].id] : []);
```
That's indeed what I had first, but it doesn't work: `selected` contains items which must be representable in the DOM.But I'll prepend `vm.connectionName` to make them unique.

cockpituous commented on src/components/vm/vmReplaceSpiceDialog.jsx:
```diff
@@ -30,42 +36,98 @@ import { fmt_to_fragments } from 'utils.jsx';
+        const onSelect = (_ev, item) => setSelected(
+            selected.includes(item)
+                ? selected.filter(id => id !== item)
+                : [...selected, item]
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240209-140217-7035aafd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

mvollmer reviewed (approved):
Danke!
